sdks/python/apache_beam/transforms/sideinputs.py

```diff
@@ -146,6 +152,63 @@ def apply(self, pcoll):
     output_type = typehints.Dict[key_type, value_type]
     return (pcoll
             | CreatePCollectionView(
-                pvalue.DictPCollectionView(pcoll.pipeline))
+                pvalue.DictPCollectionView(
+                    pcoll.pipeline,
+                    default_window_mapping_fn(pcoll.windowing.windowfn)))
             .with_input_types(input_type)
             .with_output_types(output_type))
+
+
+# Top-level function so we can identify it later.
+def _global_window_mapping_fn(w, global_window=window.GlobalWindow()):
+  return global_window
+
+
+def default_window_mapping_fn(target_window_fn):
+  if target_window_fn == window.GlobalWindows():
+    return _global_window_mapping_fn
+  else:
+    def map_via_end(source_window):
+      return list(target_window_fn.assign(
+          window.WindowFn.AssignContext(source_window.max_timestamp())))[-1]
+    return map_via_end
+
+
+class SideInputMap(object):
+  """Represents a mapping of windows to side input values."""
+
+  def __init__(self, view_class, view_options, iterable):
+    self._window_mapping_fn = view_options.get(
+        'window_mapping_fn', _global_window_mapping_fn)
+    self._view_class = view_class
+    self._view_options = view_options
+    self._iterable = iterable
+    self._cache = {}
+
+  def __getitem__(self, window):
+    if window not in self._cache:
+      target_window = self._window_mapping_fn(window)
+      self._cache[window] = self._view_class.from_iterable(
+          _FilteringIterable(self._iterable, target_window), self._view_options)
```

Comment:
Yes. I'm deferring performance improvements until we see what this looks like in the face of the Fn API. (E.g. rather than from iterable, the underlying "reader" is likely to expose a (key, window) -> iterable mapping.)